File: GranvilleOpenMP/GranvilleOpenMP/Granville.cpp
#include<iostream>
#include <chrono>
#include <omp.h>

#define A 2
#define B 100

using namespace std;

int main(int argc, char *argv[])
{
    int step;

    int set_size_S = 0;
    int set_S[B - A];

    int start_num;
    int local_current_set_size_S = 0;
    int div_sum, n, i;
    int rem; // = B % step - start_num + step;

    int nr_threads;

    chrono::high_resolution_clock::time_point t1, t2;

    set_S[0] = 1;
    set_size_S = 1;

    //Get program argument
    if (argc != 2)
    {
        cout << "Usage: program.exe N" << endl;
        cout << "Where N is the number of threads to be generated" << endl;
        return 1;
    }

    //Get number of threads
    nr_threads = atoi(argv[1]);

    if (nr_threads <= 0)
    {
        cout << "Please input a positive non null number!" << endl;
        return 2;
    }

    step = nr_threads;

    //Get initial time
    t1 = chrono::high_resolution_clock::now();

    omp_set_num_threads(nr_threads);
#pragma omp parallel private(i, n, start_num, div_sum, local_current_set_size_S, rem) shared(set_S, set_size_S)
    {
        start_num = omp_get_thread_num();
        n = start_num + A;
        rem = B % step - start_num + step;

#pragma omp critical 
        {
            local_current_set_size_S = set_size_S;
        }

        while (n < B)
        {
            div_sum = 0;
            for (i = 0; i < local_current_set_size_S; i++)
            {
                if (n % set_S[i] == 0 && set_S[i] < n)
                {
                    div_sum += set_S[i];
                }
                if (div_sum > n || set_S[i] >= n)
                {
                    break;
                }
            }

            if (div_sum <= n)
            {
                if (div_sum == n)
                {
                    cout << "Found " << n << endl;
                }

#pragma omp critical 
                {
                    set_S[set_size_S++] = n;
                }
            }

            //If it is at the last iteration it shouldn't wait
            //This is to avoid threads hanging for other threads that exited the loop already
            if (n + rem < B)
            {
#pragma omp barrier
            }

#pragma omp critical 
            {
                local_current_set_size_S = set_size_S;
            }

            if (n + rem < B)
            {
#pragma omp barrier
            }

            n += step;
        }
    }

    /* Compute the execution time*/
    t2 = chrono::high_resolution_clock::now();
    chrono::duration<double> time_span = chrono::duration_cast<chrono::duration<double>>(t2 - t1);

    cout << endl << "Total time: " << time_span.count() << " seconds." << endl;

    return 0;
}

File: Granville/Granville/Granville.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Granville
{
    public class Granville
    {
        private const int A = 2;
        private const int B = 300000;

        private volatile int step;

        private volatile Barrier barrier;

        private volatile object setSLock = new object();
        private volatile IList<int> SetS = new List<int>();

        public Granville(int numThreads)
        {
            this.step = numThreads;
            this.barrier = new Barrier(numThreads);
            SetS.Add(1);
        }

        public void IsGranville(int startNr)
        {
            int currentSetSize = 0;
            int divSum = 0;
            int i, nr;
            int rem = B % step - startNr + step;

            lock (setSLock)
            {
                currentSetSize = SetS.Count();
            }

            for (nr = startNr + A; nr < B; nr += step)
            {
                divSum = 0;
                for (i = 0; i < currentSetSize; i++)
                {
                    if (nr % SetS[i] == 0 && SetS[i] < nr)
                    {
                        divSum += SetS[i];
                    }
                    if (divSum > nr || SetS[i] >= nr)
                    {
                        break;
                    }
                }

                if (divSum <= nr)
                {
                    if (divSum == nr)
                    {
                        Console.WriteLine("Found " + nr);
                    }

                    lock (setSLock)
                    {
                        SetS.Add(nr);
                    }
                }

                //If it is at the last iteration it shouldn't wait
                //This is to avoid threads hanging for other threads that exited the loop already
                if (nr + rem < B)
                {
                    barrier.SignalAndWait();
                }

                lock (setSLock)
                {
                    currentSetSize = SetS.Count();
                }

                if (nr + rem < B)
                {
                    barrier.SignalAndWait();
                }
            }
        }
    }
}
